### DISCUSSION TITLE
Stencila VSCode extension

### DISCUSSION THREAD

**Discussion by nokome** [stencila/stencila#2134] [Ideas]
This is just to kick off a discussion about what features we might have in a VSCode extension for Stencila. Please 🙏🏽 feel free to add your ideas, and comment and/or +1 👍🏽 other peoples ideas.

# Audience

This extension will target users who are already using VSCode, and as such, are towards the "coder" end of the spectrum of Stencila users (we'll probably be developing other extensions for other platforms e.g Google Docs that target "non-coder" users). We want to provide VSCode users with easy access to Stencila's functionality particularly around,

- programmable documents i.e. control flow node types in Stencila documents e.g. `ForBlock`, `CallBlock`
- previewing rendered HTML for documents including generated plots and interactive elements
- specialized AI assistants which perform well at specific tasks involved in writing reproducible research articles
- access to functionality provided by Stencila plugins (e.g. the RO Crate that is in development) 

# Features

A starting list...

## Syntax highlighting for Stencila's Markdown extensions

Stencila has several syntax extensions for Markdown that should have highlighting e.g.

````md
::: for region in regions

```python exec
# Some python code
```

:::
````

## Previewing rendered document

Stencila has its own HTML rendering engine for Markdown (and other formats) which include interactive Web Components for the different node types in the Stencila Schema. It would be useful for users to be able to live preview this within VSCode.

There are several "preview markdown" side-by-side extensions already (e.g. [Markdown Preview Github Styling](https://github.com/mjbvz/vscode-github-markdown-preview-style) and [`quarto-vscode`](https://github.com/quarto-dev/quarto/tree/main/apps/vscode)). 

In the Stencila "writer app" we provide a similar side-by-side view (example below). What we want to do is bring the right pane into VSCode so users can edit Stencila Markdown there and have a live preview side-by-side.

![image](https://github.com/stencila/stencila/assets/1152336/93d5a9ef-4861-43d8-9446-9b185eb18999)

## Code completion for Stencila's Markdown extensions

For example, when a user types `::: for`, the extension should provide a code completion snippet.

## Autocomplete suggestions using Stencila assistants and node stores

Stencila has several task specialized AI assistants and the concept of "node stores" - places from which nodes of different types in the Stencila Schema can be stored. In addition to the builtin assistants and stores. Plugins can provide these.

We want to allow users to be able to use assistants and stores from within VSCode, most likely via an autocomplete prompt.

**Comment by GusEllerm** [comment]
Is there a description of the node terms and their 'class' for the MD format - a source of truth for highlighting / accenting terms?

**Reply by nokome** [reply]
No unfortunately there is not a reference for that yet. We a still finalizing the syntax for Stencila's extensions (another discussion thread forthcoming on that!). But that definitely needs to be prioritized and made available before the syntax highlighting work on this project can be started.
